### PR TITLE
nix: make fetching Gradle dependencies more generic

### PR DESCRIPTION
**Pull request by jakubgs**
Before what we did was essentially guess what files might exist for any given package. This approach mostly works, but not entirely. This is especially problematic when dealing with weird edge case packages like `react-native`, which you can read about here:

* https://github.com/react-native-community/discussions-and-proposals/pull/508
* https://github.com/react-native-community/discussions-and-proposals/blob/4a06fc64/proposals/0508-out-of-npm-artifacts.md#the-react-native-android-archive

In order to avoid as much the guessing aspect of fetching Gradle dependencies we are using both HTML listsings of files and `artifact-metadata.json` files that exist for more recent packages.

This way we can avoid having to add special edge cases that have been found out when working on React Native 72 upgrade in:
* https://github.com/status-im/status-mobile/pull/17062

### CODE REVIEW

**Comment by jakubgs** [comment]
Here's the difference in sizes of the Gradle dependencies derivation:

### Before
```
 > du -hsL $(nix-build --no-out-link -A pkgs.deps.gradle)
957M	/nix/store/6dahmcwyygam7whvwi3zsrgxa49i5xwp-status-mobile-maven-deps

 > find -L $(nix-build --no-out-link -A pkgs.deps.gradle) -type f | wc -l
4651
```
### After
```
 > du -hsL $(nix-build --no-out-link -A pkgs.deps.gradle)
1010M	/nix/store/0v8m8a321pp3vf01vn7inf5m3xl0rbpz-status-mobile-maven-deps

 > find -L $(nix-build --no-out-link -A pkgs.deps.gradle) -type f | wc -l
7585
```
---
Which is not horrible. Out of the new files most are the `.asc` files with GPG signatures. To be exact 1445.

This is annoying because as far as I can tell only one package actually requires the `.asc` file is `conceal-1.1.3`:
```
Could not determine the dependencies of task ':app:lintVitalReportPr'.
> Could not resolve all task dependencies for configuration ':app:prRuntimeClasspath'.
   > Could not resolve com.facebook.conceal:conceal:1.1.3.
     Required by:
         project :app > project :react-native-keychain
      > No cached version of com.facebook.conceal:conceal:1.1.3 available for offline mode.
```

File: nix/deps/gradle/url2json.sh [hunk 1]
```diff
@@ -14,6 +14,10 @@ declare -a REPOS=(
   "https://jitpack.io"
 )
 
+# These files are not necessary for the build process.
+FILENAMES_BLACKLIST='-(javadoc|runtime|gwt|headers|sources|src|tests|adapters|modular|site|bin)\.'
+FILETYPES_BLACKLIST='(pom|json|zip|module|xml|md5|sha1|sha256|sha512)$'
```
Comment (siddarthkay) [inline nix/deps/gradle/url2json.sh:19]: just curious can we also add `asc` files to Blacklist?

Comment (siddarthkay) [reply nix/deps/gradle/url2json.sh:19]: Oh my bad didn't read your description properly.

**Review by siddarthkay — APPROVED** [review]
Tested locally and it works.
My benchmarks for single runs :
- `make nix-update-gradle` ~ 18 mins
- `make run-android` ~ 13 mins

**Comment by jakubgs** [comment]
I think I can make the blacklist more strict. I believe we only need ASC files for AARs - of which there is only 32 - for builds.

**Comment by jakubgs** [comment]
That's better:
```
 > du -hsL $(nix-build --no-out-link -A pkgs.deps.gradle)
1006M	/nix/store/8g5vzlrfhihy7pv7ya6f29hr4941n9kq-status-mobile-maven-deps

 > find -L $(nix-build --no-out-link -A pkgs.deps.gradle) -type f | wc -l
4763
```
A `make nix-update-gradle` run for me takes 7 minutes.

**Review by yakimant — APPROVED** [review]

**Pull request merged**